core/embed/rust/src/ui/component/jpeg.rs

```diff
@@ -0,0 +1,98 @@
+use crate::{
+    error::Error,
+    micropython::{buffer::get_buffer, obj::Obj},
+};
+
+use crate::ui::{
+    component::{Component, Event, EventCtx, Never},
+    display,
+    geometry::{Alignment2D, Offset, Rect},
+    shape,
+    shape::Renderer,
+};
+
+pub enum ImageBuffer {
+    Object { obj: Obj },
+    Slice { data: &'static [u8] },
+}
+
+impl ImageBuffer {
+    pub fn from_object(obj: Obj) -> Result<Self, Error> {
+        if !obj.is_bytes() {
+            return Err(Error::TypeError);
+        }
+        Ok(ImageBuffer::Object { obj })
+    }
+
+    pub fn from_slice(data: &'static [u8]) -> Self {
+        ImageBuffer::Slice { data }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.data().is_empty()
+    }
+
+    pub fn data(&self) -> &[u8] {
+        match self {
+            // SAFETY: We expect no existing mutable reference. Resulting reference is
+            // discarded before returning to micropython.
+            ImageBuffer::Object { obj } => unsafe { unwrap!(get_buffer(*obj)) },
+            ImageBuffer::Slice { data } => data,
+        }
+    }
```

Comment:
I've completely refactored the image decompression process - please see this commit e364a82cef5378b09aba522874021b9272d81596.

- The logic for `ImageBuffer` has been moved to `io::BinaryData`.
- `UzlibContext` has been replaced by `ZlibInflate`, which provides stream-like decompression.
- `ZlibCache` and `JpegCache` have been updated to use `BinaryData` instead of `&[u8]`.
- A new `display::image` module for parsing image headers has been introduced, replacing the `Toif` structure and the `tjpgd` module.
- `HomeScreen`, `LockScreen`, and home screen confirmation have been updated accordingly.

FYI @TychoVrahe